File: 백준/Silver/11659. 구간 합 구하기 4/구간 합 구하기 4.cc
#include <iostream>
#include <vector>
using namespace std;

int n, m;

int main(){
    ios::sync_with_stdio(0);
    cin.tie(nullptr);
    cout.tie(nullptr);

    cin >> n >> m;
    vector<int> numbers(n), sum(n);

    for(int i = 0; i< n; i++){
        cin >> numbers[i];

        // 첫 번째 숫자는 sum 배열에 그대로 입력
        if(i == 0){
            sum[i] = numbers[i];
            continue;
        }

        // 누적합 저장
        sum[i] = numbers[i] + sum[i-1];
    }

    for(int i = 0; i< m; i++){
        int a, b;
        cin >> a >> b;

        // 1번 index부터의 합이 질문일 경우 누적합인 sum[b]값 그대로 출력
        if(a == 1){
            cout << sum[b-1] << "\n";
            continue;
        }

        // numbers[a] + .. + numbers[b] = sum[b] - sum[a-1]
        cout << sum[b-1] - sum[a-2] << "\n";
    }

    return 0;
}

File: 백준/Silver/11659. 구간 합 구하기 4/구간 합 구하기 4.java
import java.util.*;
import java.io.*;

public class Main {
    static int n, m;
    static ArrayList<Integer> numbers = new ArrayList<>();
    static ArrayList<Integer> sum = new ArrayList<>();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());
        for(int i = 0; i<n; i++){
            numbers.add(Integer.parseInt(st.nextToken()));

            // 첫 번째 숫자는 그대로 sum에 저장
            if(i == 0){
                sum.add(numbers.get(0));
                continue;
            }

            // 입력 받으며 누적합 저장
            sum.add(sum.get(i - 1) + numbers.get(i));
        }


        for(int i = 0; i<m; i++){
            st = new StringTokenizer(br.readLine());
            int from = Integer.parseInt(st.nextToken());
            int to = Integer.parseInt(st.nextToken());

            // 1번 index부터의 합이 질문일 경우 누적합인 sum[b]값 그대로 출력
            if(from == 1){
                System.out.println(sum.get(to-1));
                continue;
            }
            // numbers[a] + .. + numbers[b] = sum[b] - sum[a-1]
            System.out.println(sum.get(to - 1) - sum.get(from - 2));
        }
    }
}
